src/Featurizers/Components/NormUpdaters.h

```diff
@@ -1,298 +1,373 @@
-    // input type can only be integer or numeric
-    static_assert(Traits<T>::IsIntOrNumeric::value, "Input type to norm updater has to be integer or numerical types!");
-
-    // ----------------------------------------------------------------------
-    // |
-    // |  Public Methods
-    // |
-    // ----------------------------------------------------------------------
-
-    MaxNormUpdater(void);
-
-    void update(T input);
-
-    typename TypeSelector::MaxNormTypeSelector<T>::type commit(void);
-
-private:
-    // ----------------------------------------------------------------------
-    // |
-    // |  Private Data
-    // |
-    // ----------------------------------------------------------------------
-    typename TypeSelector::MaxNormTypeSelector<T>::type _max_norm;
-
-    // flags to check if update functions are called
-    // an error will be thrown if no update(input) is called before commit
-    // since it doesn't make sense to create an updater and commit without any update
-    bool                                                _update_flag;
-};
-
-} // Updaters
-
-// ----------------------------------------------------------------------
-// ----------------------------------------------------------------------
-// ----------------------------------------------------------------------
-// |
-// |  Implementation
-// |
-// ----------------------------------------------------------------------
-// ----------------------------------------------------------------------
-// ----------------------------------------------------------------------
-
-// ----------------------------------------------------------------------
-// |
-// |  Updaters::L1NormUpdater
-// |
-// ----------------------------------------------------------------------
-template <typename T>
-Updaters::L1NormUpdater<T>::L1NormUpdater() :
-    _l1_norm(0),
-    _update_flag(false) {
-}
-
-template <typename T>
-void Updaters::L1NormUpdater<T>::update(T input) {
-    if (!_update_flag) {
-        _update_flag = true;
-    }
-    long double diff = std::abs(static_cast<long double>(input));
-
-    // check if diff is too small comparing to l1_norm
-    if ((_l1_norm + diff == _l1_norm) && (diff != 0)) {
-        throw std::runtime_error("In l1 norm calculation, input is so small comparing to l1_norm that l1_norm is the same after long double addition!");
-    }
-    _l1_norm += diff;
-
-    // check if exceeds bound
-    // only check for l2_norm and no checks for l1_norm is because if l1_norm is out of upper bound, l2_norm is definitely out of bound
-    if (std::isinf(_l1_norm)) {
-        throw std::runtime_error("Overflows occured during calculating l1_norm! Check your data!");
-    }
-}
-
-template <typename T>
-long double Updaters::L1NormUpdater<T>::commit(void) {
-    assert(_l1_norm >= 0);
-    if (!_update_flag) {
-        throw std::runtime_error("update is not called before l1 is committed!");
-    }

```

Comment:
By removing this clause, it is possible that commit can be called when no valid input is provided. I think that this should be restored, as that is an error condition.